### PR TITLE
Merge decimal

### PR DESCRIPTION
**Pull request by srittau**

### CODE REVIEW

File: stdlib/2and3/decimal.pyi [hunk 1]
```diff
@@ -13,14 +12,21 @@ DecimalTuple = NamedTuple('DecimalTuple',
                            ('digits', Sequence[int]),  # TODO: Use Tuple[int, ...]
```
Comment (JelleZijlstra) [inline stdlib/2and3/decimal.pyi]: Why don't we use `Tuple[int, ...]`?

File: stdlib/2and3/decimal.pyi [hunk 2]
```diff
@@ -49,24 +55,37 @@ class Overflow(Inexact, Rounded): ...
...
 
-class Decimal(SupportsAbs[Decimal], SupportsFloat, SupportsInt):
-    def __init__(cls, value: Union[_Decimal, float, str, unicode,
+if sys.version_info >= (3,):
+    _SupportsRound = SupportsRound[int]
+else:
+    _SupportsRound = object
+
+class Decimal(SupportsAbs[Decimal], SupportsFloat, SupportsInt, _SupportsRound):
+    # TODO: SupportsCeil, SupportsFloor, SupportsTrunc?
```
Comment (JelleZijlstra) [inline stdlib/2and3/decimal.pyi]: Aren't these explicit base classes mostly unnecessary now because of protocols?

File: stdlib/2and3/decimal.pyi [hunk 3]
```diff
@@ -49,24 +55,37 @@ class Overflow(Inexact, Rounded): ...
...
+
+    def __init__(cls, value: Union[_Decimal, float, Text,
                                    Tuple[int, Sequence[int], int]] = ...,
                  context: Context = ...) -> None: ...
```
Comment (JelleZijlstra) [inline stdlib/2and3/decimal.pyi]: context is Optional

File: stdlib/2and3/decimal.pyi [hunk 4]
```diff
@@ -49,24 +55,37 @@ class Overflow(Inexact, Rounded): ...
...
-    def __init__(cls, value: Union[_Decimal, float, str, unicode,
+if sys.version_info >= (3,):
+    _SupportsRound = SupportsRound[int]
+else:
+    _SupportsRound = object
+
+class Decimal(SupportsAbs[Decimal], SupportsFloat, SupportsInt, _SupportsRound):
+    # TODO: SupportsCeil, SupportsFloor, SupportsTrunc?
+
+    def __init__(cls, value: Union[_Decimal, float, Text,
```
Comment (JelleZijlstra) [inline stdlib/2and3/decimal.pyi]: This is actually `__new__`, not `__init__`, at least for the Python implementation. It matters for subclassing.

File: stdlib/2and3/decimal.pyi [hunk 5]
```diff
@@ -49,24 +55,37 @@ class Overflow(Inexact, Rounded): ...
...
                                    Tuple[int, Sequence[int], int]] = ...,
                  context: Context = ...) -> None: ...
     @classmethod
     def from_float(cls, f: float) -> Decimal: ...
-    def __nonzero__(self) -> bool: ...
+    if sys.version_info >= (3,):
+        def __bool__(self) -> bool: ...
+    else:
+        def __nonzero__(self) -> bool: ...
     def __eq__(self, other: object) -> bool: ...
```
Comment (JelleZijlstra) [inline stdlib/2and3/decimal.pyi]: All these actually take an optional (and Optional) `context` argument.

File: stdlib/2and3/decimal.pyi [hunk 6]
```diff
@@ -49,24 +55,37 @@ class Overflow(Inexact, Rounded): ...
...
     def __ne__(self, other: object) -> bool: ...
     def __lt__(self, other: _ComparableNum) -> bool: ...
     def __le__(self, other: _ComparableNum) -> bool: ...
     def __gt__(self, other: _ComparableNum) -> bool: ...
     def __ge__(self, other: _ComparableNum) -> bool: ...
-    def compare(self, other: _Decimal) -> Decimal: ...
+    def compare(self, other: _Decimal, context: Optional[Context] = ...) -> Decimal: ...
     def __hash__(self) -> int: ...
     def as_tuple(self) -> DecimalTuple: ...
```
Comment (JelleZijlstra) [inline stdlib/2and3/decimal.pyi]: Missing `as_integer_ratio`, which is `def as_integer_ratio(self) -> Tuple[int, int]`. (Haven't checked yet whether it exists in Python 2.)

File: stdlib/2and3/decimal.pyi [hunk 7]
```diff
@@ -49,24 +55,37 @@ class Overflow(Inexact, Rounded): ...
...
     def to_eng_string(self, context: Context = ...) -> str: ...
```
Comment (JelleZijlstra) [inline stdlib/2and3/decimal.pyi]: context is Optional (and that's probably true for the other context arguments in this file). Also, all the dunders below down to `__float__` take a context argument.

File: stdlib/2and3/decimal.pyi [hunk 8]
```diff
@@ -99,28 +119,46 @@ class Decimal(SupportsAbs[Decimal], SupportsFloat, SupportsInt):
...
     def __complex__(self) -> complex: ...
-    def __long__(self) -> long: ...
+    if sys.version_info >= (3,):
+        def __round__(self, n=...) -> int: ...
+        def __floor__(self) -> int: ...
+        def __ceil__(self) -> int: ...
+    else:
+        def __long__(self) -> long: ...
     def fma(self, other: _Decimal, third: _Decimal, context: Context = ...) -> Decimal: ...
     def __pow__(self, other: _Decimal) -> Decimal: ...
```
Comment (JelleZijlstra) [inline stdlib/2and3/decimal.pyi]: Two more arguments: `def __pow__(self, other, modulo=None, context=None):`.

File: stdlib/2and3/decimal.pyi [hunk 9]
```diff
@@ -99,28 +119,46 @@ class Decimal(SupportsAbs[Decimal], SupportsFloat, SupportsInt):
...
     def __rpow__(self, other: int) -> Decimal: ...
     def normalize(self, context: Context = ...) -> Decimal: ...
     def quantize(self, exp: _Decimal, rounding: str = ...,
```
Comment (JelleZijlstra) [inline stdlib/2and3/decimal.pyi]: rounding is Optional (same for other arguments named rounding below)

**Comment by srittau** [comment]
I'd like to do these cleanups in a second PR (probably next week). When merging I noticed some of the problems you mentioned and was planning to review the stub thoroughly after the merge. I will also complete the missing annotations.

File: stdlib/2and3/decimal.pyi (hunk 4 above)
Comment (srittau) [reply stdlib/2and3/decimal.pyi]: This is actually the whole reason I started to work on this. This problem was spotted by `find-problems.py` in https://github.com/srittau/typeshed-completion.

**Pull request merged**